### PR TITLE
JS: Phi edge barrier guards

### PR DESCRIPTION
**Pull request by asgerf**
Sanitizer guards now sanitize specific inputs to phi nodes when the corresponding control-flow predecessor is dominated by the guard.

The test case explains the situation:
```js
function phi() {
  let x = source();

  if (something(x) && isSafe(x)) { // isSafe(x) is sanitizer guard
    // this input to the phi node for 'x' should be sanitized
  } else {
    x = null;
  }
  sink(x); // OK
}
```

[Evaluation](https://git.semmle.com/asger/dist-compare-reports/tree/phi-edge-barrier-guards-base_1574603300935) of security suite on nightly looks ok

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/dataflow/Configuration.qll [hunk 1]
```diff
@@ -363,6 +363,51 @@ private predicate barrierGuardBlocksAccessPath(BarrierGuardNode guard, boolean o
...
+  )
+}
+
+/**
+ * Holds if there is a barrier edge `pred -> succ` in `cfg` either through an explicit barrier edge
+ * or one implied by a barrier guard.
+ *
+ * Only holds for barriers that should apply to all flow labels.
+ */
+private predicate isBarrierEdge(Configuration cfg, DataFlow::Node pred, DataFlow::Node succ) {
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/dataflow/Configuration.qll]: Can we avoid the ambiguity with the `Configuration::isBarrierEdge` name? The current naming scheme is begging  for a mistake somewhere. 

For inspiration: we already use the `Internal` suffix for a similar predicate.

```
  private predicate isBarrierGuardInternal(BarrierGuardNode guard) {
    isBarrierGuard(guard)
    or
    guard.(AdditionalBarrierGuardNode).appliesTo(this)
  }
```

(remember to also rename `isLabeledBarrierEdge` below if you end up renaming this predicate)

File: javascript/ql/test/library-tests/TaintTracking/sanitizer-guards.js [hunk 2]
```diff
@@ -52,3 +52,14 @@ function reflective() {
     sink(x); // OK
   }
 }
+
+function phi() {
```
Comment (esbena) [inline javascript/ql/test/library-tests/TaintTracking/sanitizer-guards.js]: Can we add this test case as well?
```js
function phi2() {
  let x = source();

  if (somethingThatProbablyMakesItSafe(x) || isSafe(x)) {
    // this input to the phi node for 'x' should be sanitized, but we can not prove that yet
  } else {
    x = null;
  }
  sink(x); // OK, but still flagged
}
```

File: javascript/ql/src/semmle/javascript/dataflow/Configuration.qll (hunk 1 above)
Comment (asgerf) [reply javascript/ql/src/semmle/javascript/dataflow/Configuration.qll]: Hm, it seems to be consistent with the top-level `isSource` and `isSink` predicates and I don't believe they've caused any problems. The "internal" naming was mainly meant to indicate that it's used locally as a helper in another predicate and isn't a final input to the data flow analysis.

File: javascript/ql/test/library-tests/TaintTracking/sanitizer-guards.js (hunk 2 above)
Comment (asgerf) [reply javascript/ql/test/library-tests/TaintTracking/sanitizer-guards.js]: Is `somethingThatProbablyMakesItSafe` supposed to be added as a guard?

The comment is wrong when it says "we can't prove this **yet**" if that's not actually a guard.

File: javascript/ql/src/semmle/javascript/dataflow/Configuration.qll (hunk 1 above)
Comment (esbena) [reply javascript/ql/src/semmle/javascript/dataflow/Configuration.qll]: :+1:

File: javascript/ql/test/library-tests/TaintTracking/sanitizer-guards.js (hunk 2 above)
Comment (esbena) [reply javascript/ql/test/library-tests/TaintTracking/sanitizer-guards.js]: The main point of the new test case is to ensure that we have a negative test for the dominance part of your changes.

`somethingThatProbablyMakesItSafe` is supposed to be a sanitizing call that we for one reason or another do not detect as such (missing dataflow for the definition of `somethingThatProbablyMakesItSafe`, missing recognition of `somethingThatProbablyMakesItSafe` as a sanitizing function, ...).

Comment (asgerf) [reply javascript/ql/test/library-tests/TaintTracking/sanitizer-guards.js]: 👍 Good point about adding a negative case.

File: javascript/ql/test/library-tests/TaintTracking/sanitizer-guards.js [hunk 3]
```diff
@@ -63,3 +63,14 @@ function phi() {
...
+
+function phi2() {
+  let x = source();
+
+  if (something(x) || isSafe(x)) {
+    // this input to the phi node for 'x' is not fully sanitized
+  } else {
+    x = null;
+  }
+  sink(x); // OK
```
Comment (esbena) [inline javascript/ql/test/library-tests/TaintTracking/sanitizer-guards.js]: This should be `// OK, but still flagged` or something like that.

Comment (asgerf) [reply javascript/ql/test/library-tests/TaintTracking/sanitizer-guards.js]: Copy pasta. Changed to `// NOT OK`.

**Comment by asgerf** [comment]
[Evaluation on big-apps](https://git.semmle.com/asger/dist-compare-reports/tree/profiling-js-asger.northeurope.cloudapp.azure.com_1575876787377) was cut short for unexplained reasons, but here's what I've salvaged.

The run on gecko-dev ended mysteriously at the 4th iteration of `Paths::PathExpr::getValue_dispred` which doesn't seem like something that this PR would cause to blow up:
```
[2019-12-08 22:00:22] (183s) Empty delta for NodeJS::JoinedPath#class#f_delta (order for disjuncts: delta=100).
[2019-12-08 22:00:22] (183s) Empty delta for Paths::ConcatPath#class#f_delta (order for disjuncts: delta=1000).
[2019-12-08 22:00:22] (183s) Empty delta for Paths::PathExpr#class#f_delta (order for disjuncts: delta=100).
[2019-12-08 22:00:22] (183s) Starting to evaluate predicate Paths::PathExpr::getValue_dispred#ff#cur_delta/2[4]@95adc9 (iteration 4)
```

**Comment by max-schaefer** [comment]
Could you redo the eval anyway? I know it's frustrating, but these changes can be subtle, so I'd rather err on the side of caution.

**Comment by asgerf** [comment]
[big-apps evaluation](https://git.semmle.com/asger/dist-compare-reports/tree/phi-edge-barrier-guards_1576154858548) done. There's an 8% slowdown on one project - I'll retry that project with DPMs.

**Comment by asgerf** [comment]
[Re-evaluation](https://git.semmle.com/asger/dist-compare-reports/tree/phi-edge-barrier-guards_1576171562757) of A2Z with DPMs show a ~1% increase in tuple counts and ~2% increase in DPM. The wall clock time was +6% but it's not clear if that's another fluke or poorly calibrated DPM weights.

**Comment by max-schaefer** [comment]
OK, thanks, that's good enough for me.

**Comment by max-schaefer** [comment]
Needs a merge/rebase due to conflict on change note.

**Comment by asgerf** [comment]
@esbena any more comments? You're still listed as requesting changes.

**Comment by esbena** [comment]
Nope. My two requests have been addressed.

**Pull request merged**